animations/shared-element-transitions/index.md

```diff
@@ -0,0 +1,684 @@
+you can't animate a Hero to a Dialog.
+For further developments (and a possible workaround), [watch this
+issue.](https://github.com/flutter/flutter/issues/10667)
+</aside>
+
+Hero animation code has the following structure:
+
+1. Define a starting Hero object, referred to as the _source
+   Hero_. The Hero specifies its location,
+   size, graphical representation (typically an image), and an
+   identifying tag.
+1. Define an ending Hero object, referred to as the _destination Hero_.
+   This Hero also specifies its location, size, graphical
+   representation, and the same tag as the source Hero.
+   It's <strong>essential that both Hero objects are created with
+   the same tag</strong>, typically an object that represents the
+   underlying data. For best results, the Heroes should have
+   virtually identical widget trees.
+1. (Optional) Place any Hero object that responds to user input,
+   such as a tap, in an Inkwell.
+1. Specify a route that contains the destination Hero.
+   The destination route defines its widget tree at the end
+   of the transition.
+1. Trigger the animation by pushing the route on the Navigator's stack.
+   Popping a route from the Navigator’s stack triggers a return
+   transition.
+
+Flutter calculates the tweens required to transition the
+starting Hero to the ending Hero (interpolating size and
+position changes), and performs the animation in an overlay.
+
+The next section describes Flutter's process in greater detail.
+
+## Behind the scenes
+
+The source route contains the Hero inside an Inkwell.
+What happens when the InkWell receives a tap gesture?
+The following describes how Flutter performs the transition.
+
+<img src="images/hero-transition-0.png" alt="Before the transition the source hero appears in the source route">
+
+Before transition, the source Hero waits in the source route's widget
+tree. The destination route does not yet exist, and the overlay
+is blank and disabled.
+
+---
+
+<img src="images/hero-transition-1.png" alt="Transition begins and Destination Hero appears in the overlay">
+
+Tapping the Hero's InkWell triggers the animation, and the
+following occurs:
+
+* Offscreen, Flutter calculates the transition (including
+  position and size changes) from beginning to end,
+  moving the Hero along a curved path as described in the
+  Material motion spec. Flutter saves the size and location
+  of the Hero at the beginning and end of the transition.
+
+* Flutter places the destination Hero in the overlay, at the
+  same location and size as the _source_ Hero.
+
+* Flutter enables the overlay, obscuring the current route.
+
+* Flutter creates the destination route, placing it at the
+  top of the Navigator's stack (above the source route),
+  but it's obscured by the overlay.
+
+* The destination route won't contain the destination Hero
+  until the transition is complete.
+
+---
+
+<img src="images/hero-transition-2.png" alt="The Hero flies in the overlay to its final position and size">
```

Comment:
This diagram isn't really correct. During the Hero animation both routes are visible although neither one contains its Hero. While the Hero is flying the destination route is animating into place (using whatever route transition animation was defined for the destination route).

It might help to diagram a few widgets (like in light gray) on the source and destination routes.